Upload Dynatest survey feature

s203860 opened the pull request:
- [x] check quickly that the javadoc are still up-to-date
- [x] TODOs in the code
- [x] check that this is giving the same results as before
- [x] check with big files
- [x] Fix pipeline

s203860 committed:
Upload: create controller, service, queue, redis
- Controller: receive files
- Queue: unzip received files
- Service: Interact with db

Seb-sti1 committed:
Upload: add password auth

Seb-sti1 commented on backend/src/app.module.ts:
```diff
@@ -41,14 +45,23 @@ const database = (config: any, name: string) => {
     }),
+    BullModule.forRoot({
+      redis: {
+        host: 'se2-d.compute.dtu.dk',
+        port: 6379,
```
use env var (when #164 merged)

liulw2022 committed:
Migrate data importation process from DynatestData repo
- Extract survey path from RSP
- Extract data from XMLs
- Extract road surface images and rotate them
- Extract dashcam images
- Save data to database

liulw2022 commented:
I couldn't find a way for the database_function.js file to call 'upload.service.ts' and get the return value. I've tried some methods like compiling the TS file to JS, but none have worked. The only solution I haven't tried is to convert the entire 'database_function' to TypeScript, but I'm not sure how to proceed from here. 

By the way, there is also a bug when uploading large datasets, specifically 'Dynatest Data' exceeding 1GB. During the extraction process, it appears to get stuck. I'm uncertain whether this is due to my computer's performance limitations or if the chosen decompression method is not efficient enough.

I would greatly appreciate it if you could take a look when you have time. Thank you very much！ 
@Seb-sti1

Seb-sti1 commented:
> I couldn't find a way for the database_function.js file to call 'upload.service.ts' and get the return value. I've tried some methods like compiling the TS file to JS, but none have worked. The only solution I haven't tried is to convert the entire 'database_function' to TypeScript, but I'm not sure how to proceed from here.

Hey, so yeah `database_function.js` needs to be converted to typescript. I have just done it (at first i wanted to give an example but to make it clear i needed to do another function... and so on until now... and then the file was converted). I will do another comment tomorrow morning to describe which function correspond to which function from `database_function.js` (i splited the function across several files to make it easier to read) but most of the function didn't change.

There is still some work to do (in particular, check quickly that the javadoc are still up-to-date, see TODO in the code and check that this is giving the same results as before). Then we will probably merge and open a new MR to return status to the frontend (so that the user knows the progress of the upload) and another one for valhalla.

**To continue, you need to do `git fetch --all ; git reset --hard origin/Upload_test` (or delete your local branch and re-checkout from the origin branch)** and add the following line in the .env in the backend:

_See WhatsApp for the full new .env_

Then to upload a zip file you can start the frontend and use the "upload" button on the main page (the password ask is the one setted in the `.env` by `UPLOAD_PASSWORD`)

![image](https://github.com/DTU-SE-Group-D/LiRA-Viz/assets/65665540/7752ff7c-26c7-4e47-9336-093325b661f5)


Let me know if you have any question regarding anything :-)

> By the way, there is also a bug when uploading large datasets, specifically 'Dynatest Data' exceeding 1GB. During the extraction process, it appears to get stuck. I'm uncertain whether this is due to my computer's performance limitations or if the chosen decompression method is not efficient enough.

Yeah was testing with MFV Ballerup 2023.zip (~133Mo), might need to alter some parameters to accept big file (that another thing todo)...

Seb-sti1 commented:
To make the importation process work for more than one zip file at the same time, I was forced to put the array of directories/files as arguments.
To keep the function easy to read, I created a new types (in `upload.ts`)

```typescript
export interface SurveyStructure {
  // our own survey id
  fk_survey_id?: string;
  // the path to the directory
  directory: string;
  // the path to the RSP file
  RSP: string;
  // the geometry of the section
  geometry?: LatLng[];
  // the path to the HDC folder
  HDC: string;
  // the paths to the XML files
  XMLs: string[];
  // the camera directories
  Cams: string[];
  // the available modules
  available_modules: boolean[];
}

export interface SurveyData {
  // data from the XML
  /** the type of measurement, see MeasurementType enum (in models.ts) for description of numbers. */
  type_index: number;
  /**  the value for the measurement, can mean different things dependent on type of measurement */
  value: number;
  /** the distance from the beginning of the survey */
  distance_survey: number;
  /** the date in format: year/month/day hour:minute:second.millisecond. */
  timestamp: string;
  // data from Valhalla
  /** the automatically generated uuid from the database for the corresponding way entry in the way table. */
  fk_way_id?: string;
  /** the distance from the beginning of the way */
  distance_way?: number;
  /** GPS point */
  position?: LatLng;
}

export interface SurveyImage {
  // data from the XML
  /** The type of image, for example, 'DashCamera'. */
  type: string;
  /** The file path where the image is stored. */
  image_path: string;
  /** The distance from the beginning of the survey at which the image was captured. */
  distance_survey: number;
  /** The timestamp of the image capture in the format: year/month/day hour:minute:second.millisecond */
  timestamp: string;
  // data from Valhalla
  fk_way_id?: string;
  distance_way?: number;
}
```

The `SurveyStructure` type basically represents one survey (so one element of each array).


| name in database_function.js       | name and new file                             | modifications                                                                                                                                                               |
|------------------------------------|-----------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| count_all_cracks                   | count_all_cracks in xml.ts                    | minor typescript conversion                                                                                                                                                 |
| count_cracks                       | count_cracks in xml.ts                        | minor typescript conversion                                                                                                                                                 |
| count_all_potholes                 | count_all_potholes in xml.ts                  | minor typescript conversion                                                                                                                                                 |
| count_potholes                     | count_potholes in xml.ts                      | minor typescript conversion                                                                                                                                                 |
| count_sealed_cracks                | count_sealed_cracks in xml.ts                 | minor typescript conversion                                                                                                                                                 |
| count_all_sealed_cracks            | count_all_sealed_cracks in xml.ts             | minor typescript conversion                                                                                                                                                 |
| count_mmo                          | count_mmo in xml.ts                           | minor typescript conversion                                                                                                                                                 |
| count_all_mmo                      | count_all_mmo in xml.ts                       | minor typescript conversion                                                                                                                                                 |
| get_roughness                      | get_roughness in xml.ts                       | minor typescript conversion                                                                                                                                                 |
| check_for_HDC_and_RSP              | find_surveys in upload.ts                     | Merged with sort_directories and findAllCamFolders.<br/> I also simplified the function knowing the zip file constraint to the user                                                  |
| sort_directories                   | find_surveys in upload.ts                     | Merged with check_for_HDC_and_RSP and findAllCamFolders.                                                                                                                    |
| findAllCamFolders                  | find_surveys in upload.ts                     | Merged with check_for_HDC_and_RSP and sort_directories.                                                                                                                     |
| get_start                          | get_start in xml.ts                           | minor typescript conversion                                                                                                                                                 |
| get_end                            | not migrated (unused)                         |                                                                                                                                                                             |
| get_length                         | not migrated (unused)                         |                                                                                                                                                                             |
| get_date                           | get_date in xml.ts                            | minor typescript conversion                                                                                                                                                 |
| get_surveyID                       | get_surveyID in xml.ts                        | minor typescript conversion                                                                                                                                                 |
| getStartAndEndCoordinatesForAllXML | not migrated (unused)                         |                                                                                                                                                                             |
| extractCoordinatesFromRSP          | extractCoordinatesFromRSP in upload.ts        | minor typescript conversion                                                                                                                                                 |
| whats_available                    | whats_available in xml.ts                     | typescript conversion + small optimization to make it easier to add new module detection (remove code duplication)                                                          |
| print_whats_available              | print_whats_available in utils.ts             | minor typescript conversion                                                                                                                                                 |
| get_first_xml_file                 | not migrated (unused anymore)                 |                                                                                                                                                                             |
| index_xml_files_in_HDC_directory   | index_xml_files_in_HDC_directory in upload.ts | minor typescript conversion                                                                                                                                                 |
| upload_measurements_data           | extract_measurements_data in upload.ts        | typescript conversion + prepare use of valhalla (return data instead of directly uploading it to the db) + small optimization to make it easier to add new module detection |
| upload_survey_data                 | handleFileProcessing in file.processor.ts     | merged with the big function                                                                                                                                                |
| upload_image_data                  | extract_road_image_data in upload.ts          | typescript conversion + prepare use of valhalla (return data instead of directly uploading it to the db)                                                                    |
| upload_DashCamera_data             | extract_dashcam_image_data in upload.ts       | typescript conversion + prepare use of valhalla (return data instead of directly uploading it to the db)                                                                    |
| the_big_function                   | handleFileProcessing in file.processor.ts     | merged with upload_survey_data. Add a for loop to extract data for all detected surveys                                                                                     |

Seb-sti1 commented on backend/src/upload/upload.ts:
```diff
@@ -65,6 +65,10 @@ export interface SurveyStructure {
   Cams: string[];
   // the available modules
   available_modules: boolean[];
+  // the survey id of the dynatest
+  dynatestId?: number;
+  // the timestamp of the survey
+  surveyDate?: string;
```
Can't they be non-optional?

Seb-sti1 replied:
Also there might some places where these values can be used instead of recomputing them using `get_surveyID` or `surveyDate`

liulw2022 replied:
> Also there might some places where these values can be used instead of recomputing them using `get_surveyID` or `surveyDate`

Now the  `get_surveyID` or `surveyDate` only be used once in the `find_survey`.

Seb-sti1 commented on backend/src/upload/upload.ts:
```diff
@@ -147,13 +151,21 @@ export function find_surveys(
+      surveyDirectory + '/HDC',
+    )[0];
+    const dynatestId = get_surveyID(firstXML, debug);
+    const surveyDate = get_date(firstXML, debug);
```
```suggestion
    const xmls = index_xml_files_in_HDC_directory(
      surveyDirectory + '/HDC',
    );
    const dynatestId = get_surveyID(xmls[0], debug);
    const surveyDate = get_date(xmls[0], debug);
```
to avoid recomputing `index_xml_files_in_HDC_directory`

Seb-sti1 commented on backend/src/upload/upload.ts:
```diff
@@ -147,13 +151,21 @@ export function find_surveys(
       directory: surveyDirectory,
       RSP: surveyDirectory + '.RSP',
       HDC: surveyDirectory + '/HDC',
       XMLs: index_xml_files_in_HDC_directory(surveyDirectory + '/HDC'),
```
```suggestion
      XMLs: xmls,
```

Seb-sti1 commented on backend/src/upload/file.processor.ts:
```diff
@@ -68,24 +92,25 @@ export class FileProcessor {
-        let surveyDate = get_date(surveys[i].XMLs[0], debug);
+        // put the following two variable in the SurveyStructure
+        // let dynatestId = get_surveyID(surveys[i].XMLs[0], debug);
+        // let surveyDate = get_date(surveys[i].XMLs[0], debug);
```
```suggestion
```
can be removed now :+1:

Seb-sti1 commented on backend/src/upload/file.processor.ts:
```diff
@@ -68,24 +92,25 @@ export class FileProcessor {
         surveys[i].geometry = extractCoordinatesFromRSP(surveys[i].RSP);
 
         // upload the survey data and get the id back
         surveys[i].fk_survey_id = await this.service.db_insert_survey_data(
```
Maybe `db_insert_survey_data` can be refactored now to take the `surveys[i]` object instead of some of the properties

liulw2022 replied:
Yes, The ```db_insert_survey_data``` function has been refactored in the latest update.

Seb-sti1 reviewed (changes requested):
@liulw2022 nice work! Didn't test it locally for now (will do that as soon as possible).

_Btw the comments aren't specifically for you, anyone of the three of us can try to resolve them when possible_

Seb-sti1 commented on backend/src/upload/xml.ts:
```diff
@@ -0,0 +1,408 @@
+ * @returns the total count of cracks for all xml files in directory as a number.
+ * @author Vejlgaard
+ */
+export function count_all_cracks(directory: string, debug: boolean): number {
+  let total_count = 0;
+  readdirSync(directory).forEach((file) => {
+    if (file.split('.').pop() == 'xml') {
+      const words = readFileSync(directory + '/' + file).toString();
+      const count = (words.match(/CrackID/g) || []).length / 2;
```
@s203860 most of the `count_all_something` are unused, can they be removed?

sungamu replied:
yes, but i hurts :(

liulw2022 commented:
Hey guys,  I've encountered an issue while debugging the upload process for our measurement table. Specifically, I ran into an error related to uploading the position data.

Upon inspecting our prodb database, I noticed that two new columns – latitude and longitude – have been added to the measurement table. I'd like to clarify whether these coordinates are intended to represent some point of each XML file, or are they supposed to represent the starting point of the entire survey or other stuff? Thanks 
@Seb-sti1 @s203860

liulw2022 commented:
Hey guys, the TODOs in our code are almost done, and the image stuff – rotation and uploads – seems to be working fine. But, I've got a couple of things I'm still wondering about.

First up, in `file.processor.ts`, we've got this TODO about splitting the process for each survey instead of doing the whole zip file at once. I think our for loop is already handling this, but I'm not 100% sure if there's anything else we need to do there.  @Seb-sti1 

Also, there are some XML-related TODOs about the conditions in `xml.ts`  that I'm not too sure about. Could you check them out when you have a moment? @s203860

Seb-sti1 commented on backend/src/upload/upload.service.ts:
```diff
@@ -35,8 +34,6 @@ export class UploadService {
         fk_way_id: data.fk_way_id,
         distance_survey: data.distance_survey,
-        latitude: data.position.lat,
-        longitude: data.position.lng,
```
this will be needed with valhalla

Seb-sti1 commented:
> Hey guys, the TODOs in our code are almost done, and the image stuff – rotation and uploads – seems to be working fine.
Very nice!

> First up, in file.processor.ts, we've got this TODO about splitting the process for each survey instead of doing the whole zip file at once. I think our for loop is already handling this, but I'm not 100% sure if there's anything else we need to do there. @Seb-sti1

Yes the `the_big_function` was capable of handling multiple survey at once. The equivalent function `handleFileProcessing` too. The interesting thing is that with a nest queue there is no reason here to make a process do multiple surveys: for instance, instead of having one process doing X surveys, we could have X process doing 1 survey. The big difference with the second version is that the server would probably do them in parallel, so overall it would take less time to import the X surveys.
That being said this is not really priority now so you can ignore this TODO and will see later if we have time to do it.

> there are some XML-related TODOs

Regarding the XML, i'm not sure if there is a way to test those (we would need to have a survey containing those indicators). I added the TODOs because i needed (because typescript) to change the condition and i was unsure that i would still work...

Seb-sti1 commented on backend/package.json:
```diff
@@ -21,28 +21,38 @@
     "doc": "compodoc -p tsconfig.doc.json"
   },
   "dependencies": {
+    "@img/sharp-win32-x64": "^0.33.0",
```
@liulw2022 Is it necessary? (it won't work on linux/macos)

Seb-sti1 replied:
But the `npm i` works perfectly fine with only `"sharp": "^0.33.0"`

Seb-sti1 replied:
The images are correctly rotated on Linux with only "sharp": "^0.33.0". Let me know if it works on windows too

liulw2022 replied:
> @liulw2022 Is it necessary? (it won't work on linux/macos)

I do need this when I'm running it  on my computer or it keeps getting an error, but I guess we can remove it? Because it's ultimately running on the server

Seb-sti1 replied:
Yes but i don't know if Ekkart will test it on windows or linux...

Seb-sti1 replied:
I just had a `optionalDependencies` so it should work on both linux and windows, could you try ?

liulw2022 replied:
YES! Right now it perfectly works.

Seb-sti1 commented:
Right now i will work on the report, but later today i will help finish some TODOs. Hopefully it can be merged today or tomorrow morning

Seb-sti1 commented:
Very good news: it works with every dataset!!!!!

Seb-sti1 commented on backend/src/upload/file.processor.ts:
```diff
@@ -0,0 +1,181 @@
+  }
+
+  async handleUnzippedContent(tempUnzipPath, originalFilePath) {
+    // TODO: is it useful to move out of the temp dir? can't we just import the data even if the dir is in a temp dir?
```
@liulw2022 do we need to move the folder (containing the survey) out of the temp dir? Can't we just import the data from the temp dir? (it is very slow to move big files)

liulw2022 replied:
I'm using this temporary folder to solve the issue of having folders with the same name after unzipping. I tried other methods, but they all ended up merging folders with the same name. One approach I attempted was appending a timestamp if a folder with the same name was detected, but this led to nested folders like `originalname_timestamp/originalname`, which wasn't ideal. It seems the main issue is that the function has to unzip before it can check for folders with the same name, and I'm not sure if there's a better solution.

In practice, moving files within the same directory seems quite fast. Although, we might not even need to worry about that since the folders get deleted afterward.

liulw2022 replied:
Sorry, I realized I misunderstood. The latest upload should have corrected it.

Seb-sti1 reviewed (commented):
I'll eat, fix the pipeline and review one last time and this MR should be done :partying_face: 

(EDIT: fix the pipeline, eat and then review)

Seb-sti1 committed:
Fix app controller test

Seb-sti1 reviewed (commented):
Other than the two questions, everything is good for me

liulw2022 reviewed (approved):
Excellent job! Thanks so much for your hard work

liulw2022 reviewed (approved):
yeahhh! I guess we can upload the final database very soon
